### PR TITLE
Removing uniqueness constraints on tables table

### PR DESCRIPTION
**Pull request by agrawaldevesh**
Summary: The tables table was inconsistent wrt to the schema resulting
from the migrations and the model schema. The migrations said there was
uniqueness constraint on table_name, but the model schema said there
should be one on (database_id, table_name)

I hit this bug when trying to add the same table via two different
databases in the UI. It refused me and it wasn't immediately apparent
what was the cause (since the code referred to the uniquenes
constraint). After a bunch of debugging I narrowed down to this
"missing" migration script: That migrates from the uniq(table_name) to
uniq(database_id, schema, table_name) and fixes both the model and
migration script

And as an aside, also adds ctags file to .gitignore

### CODE REVIEW

**Comment by agrawaldevesh** [comment]
This is an important bugfix when you want to add the same table via multiple schemas in vertica or when you have a presto table replicated across multiple different databases. Logically speaking a table is only unique within the <database, schema>.

File: superset/migrations/alembic.ini [hunk 1]
```diff
@@ -25,7 +25,7 @@ handlers = console
 qualname =
 
 [logger_sqlalchemy]
-level = WARN
```
Comment (agrawaldevesh) [inline superset/migrations/alembic.ini]: I like having more insight into what alembic is doing. I think that level of verbosity is a good idea for such risky operations like database upgrading.

**Comment by mistercrunch** [comment]
There's a bigger question here as to whether we want any sort of unique constraint on this. You could have different owners that want to have diverging definition on top of the same table. There are pro/cons here...

Contrainst in multi-engine alembic is a shitshow, we have functions somewhere to help with this. Check out `superset.utils.core.*constraint.*`

**Comment by agrawaldevesh** [comment]
I kind of agree with you: Currently, the SQL definition aspect is tied to a table, and that too the SQL itself is not shown on the CRUD Ui so itsn't apparent then 'which version of the table to choose'. I agree that ideally we want no uniqueness.

I am happy to change this to no uniqueness constraint whatsoever and also exposing the SQL definition on the CRUD UI.

What would you suggest I do here: Keep the <database, schema, table> uniqueness but do the refactoring for the constraint stuff, OR should we remove constraints all together and/or expose the SQL definition on the CRUD ?

Basically, the current scenario of uniqueness of table-name alone is broken and I am interested in getting past that pain point first.

File: superset/migrations/versions/8d49a37823bf_remove_tables_uniqueness.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    return bind and bind.dialect and bind.dialect.name and bind.dialect.name.startswith('sqlite')
+
+def upgrade():
+    bind = op.get_bind()
+    insp = sa.engine.reflection.Inspector.from_engine(bind)
+    constraints = insp.get_unique_constraints('tables')
+    print(f'Printing the existing uniq constraints in table {constraints}')
+    if not constraints:
+        constraints = []
+    if is_sqlite():
```
Comment (agrawaldevesh) [inline superset/migrations/versions/8d49a37823bf_remove_tables_uniqueness.py]: This hacking of is-sqlite checking was necessary to force it to drop the unique constraint. In the absence of this (even with batch_alter_table, recreate=always) it wasn't really creating a new copy of the table with the uniq constraint removed. 

This is a bit hacky for sure, but it works for mysql, sqlite and postgresql. Further, it only pays the recreate cost when needed. 

I think it would be a bit premature to refactor this and move into utils, but I would be happy to put that there just as well.

File: superset/models/helpers.py [hunk 3]
```diff
@@ -97,10 +97,13 @@ def formatter(c):
 
     @classmethod
     def import_from_dict(cls, session, dict_rep, parent=None,
-                         recursive=True, sync=[]):
+                         recursive=True, sync=[], respect_id=True):
```
Comment (agrawaldevesh) [inline superset/models/helpers.py]: I am not sure of the changes in this entire file: They are working around test failures but I think they are hack and don't help. I will revert them once we have agreed about what to do about the failing tests

File: superset/connectors/sqla/models.py [hunk 4]
```diff
@@ -251,7 +251,6 @@ class SqlaTable(Model, BaseDatasource):
     owner_class = security_manager.user_model
 
     __tablename__ = 'tables'
-    __table_args__ = (UniqueConstraint('database_id', 'table_name'),)
```
Comment (agrawaldevesh) [inline superset/connectors/sqla/models.py]: Removed any table constraints as suggested by @mistercrunch

File: superset/migrations/versions/8d49a37823bf_remove_tables_uniqueness.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        # We don't want to preserve the existing table_args for the tables table
+        with op.batch_alter_table('tables', copy_from=existing_table, recreate="always") as batch_op:
+            pass
+    else:
+        print(f'Doing non sqlite transformation')
+        for c in constraints:
+            name = c.get('name', None)
+            if name:
+                try:
+                    print(f'Dropping the constraint by name {name} ie {c}')
```
Comment (agrawaldevesh) [inline superset/migrations/versions/8d49a37823bf_remove_tables_uniqueness.py]: Will remove the debug statements before commiting

File: superset/migrations/versions/8d49a37823bf_remove_tables_uniqueness.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            autoload=True,
+            autoload_with=op.get_bind())
+        existing_table.constraints = set([c for c in existing_table.constraints if not is_unique_constraint(c)])
+        print(f'Doing sqlite bulk transformation with constraints {pprint.pformat(existing_table.constraints)}')
+        # We don't want to preserve the existing table_args for the tables table
+        with op.batch_alter_table('tables', copy_from=existing_table, recreate="always") as batch_op:
+            pass
+    else:
+        print(f'Doing non sqlite transformation')
+        for c in constraints:
```
Comment (agrawaldevesh) [inline superset/migrations/versions/8d49a37823bf_remove_tables_uniqueness.py]: The reason I am removing all the table unique constraints is that in certain databases like postgresql, the uniq-constraint is on <database-id, schema, table> (caused by an old migration b4456560d4f3_change_table_unique_constraint.py), but that very same migration errors out in mysql, wherein the constraint is on <table-name>. Whereas the constraint is on <database-id, table_name> in sqlite. 

As you said, alembic migrations are a royal shitshow and frankly I think this is not scaleable :(. 

Anyway, at least the DB migration aspect of this diff works for all database types.

**Review by agrawaldevesh — COMMENTED** [review]
@mistercrunch, I need your help fixing the failing unit tests (example https://travis-ci.org/agrawaldevesh/incubator-superset/jobs/482223605). 

Do people really use the superset CLI's import_datasources and export_datasources that writes the sqla Databses and Druid cluster configuration to a YAML file and back ?

The failing unit tests (dict_import_export_tests.py's test_import_table_override_append) test this feature that one can export a YAML file, change some stuff and re-import it: Which should replace the existing configuration. 

The problem is that the "id" field is not exposed in the YAML file, and so it determines uniqueness using the table-name for the SqlaTable. That's exactly the behavior we are trying to change. So I don't know how to fix these tests without actually having SqlaTable have unique fields :( :(. 

So is this behavior of YAML import/export really used ?

**Comment by agrawaldevesh** [comment]
Ping @mistercrunch ... The test is failing because of the unit test issue with dict_import_export_tests.py. I need your advise on what to do here. I am not sure how to make this change without breaking/changing existing yaml save/restore functionality. At the same time, I think I need this fix since it prevents me from adding the different presto/vertica table names that have same name but diff schema.

**Comment by mistercrunch** [comment]
Personally I feel like the right thing to do is to remove the unique constraint.

For the sake of moving forward in your environment, you can probably just drop the constraint directly against the database.

The import/export logic is not great. It's been buggy since it was contributed. Let me see if I can help.

**Comment by mistercrunch** [comment]
Oh btw another ruse to go around the fact that some contraints will never get deleted by alembic on some database engines, is to go on the original migration that created the constraint in the first place. While this doesn't fix existing environments, it fixes the issue for brand new installations.

**Comment by mistercrunch** [comment]
Oh let me try to explain again. Since there are no easy way to clean up constraints on some database engines, we can go and edit the original database migration that created the constraint in the first place and edit it to remove the line of code that creates the constraint.

This does not fix existing installation, but it prevents new installation from creating the constraint in the first place.

**Comment by agrawaldevesh** [comment]
@mistercrunch, Thanks for the suggestion of editing the original migration file directly. I feel its a bit hacky :) It also goes against the alembic approach. 

I feel the problem isn't related to sqlite itself: Meaning, yes the alembic code needs to be different for sqlite vs non sqlite codepaths but that can be refactored away. 

I think the migration code I have here is fine, except that it breaks some obscure feature of export/import. So I am wondering if you think we should first try to fix that and if that does not work, then fall back to modifying the older migration scripts ?

What do you think ? Or I can always give up and hack it to what you are suggesting :)

File: superset/connectors/sqla/models.py (hunk 4 above)
Comment (john-bodley) [reply superset/connectors/sqla/models.py]: @agrawaldevesh I'm not sure why we should be removing this uniqueness constraint. I've actually proposed having a more restrictive uniqueness constraint per https://github.com/apache/incubator-superset/pull/5449.

**Comment by agrawaldevesh** [comment]
@mistercrunch, the PR is ready to be merged in. As we discussed, it simply expands the uniqueness constraints as also suggested by @john-bodley. It causes no other backward incompat issues. 

@kristw, I am not sure what the tag risk-db-migration means. This diff adds a new migration but it does not require a manual step, just the usual superset db upgrade command.

**Comment by kristw** [comment]
@agrawaldevesh We flag any PR with db-migration to be aware of them before including in release and testing on staging/production environment.
